### PR TITLE
[WIP] mcs augment

### PR DESCRIPTION
**Pull request by lsg1213**
일단 gpu연산이 안 되기도 하고.. 여러모로 현재 상태로는 쓰기에 좀 부족합니다. 코드 리뷰 부탁드립니다.

### CODE REVIEW

File: transforms.py [hunk 1]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+
+def stab(matrix, num_channel, theta):
+    # matrix: (batch, freq, chan, chan)
+    nx = tf.newaxis
+    dd = tf.constant([1e-6, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1], dtype=matrix.dtype)
+    for i in range(6):
+        mask = 1 - tf.cast(is_invertible(matrix), matrix.dtype)
+        matrix = matrix + (mask * dd[i])[..., nx, nx] * tf.eye(matrix.shape[-1], dtype=matrix.dtype)[nx, nx,...]
+
+    return matrix
```
Comment (ParkJinHyeock) [inline transforms.py:189]: 여기서 num_channel이랑 theta의 쓰임이 없는 것 같습니다. 그리고 코드가 음 약간 헷갈리는데, 어제 말하셨던 
여러번 곱해지면서 nan이 발생하는 상황때문에 만드신건가요..? invertible하지 않으면 constant를 키워가면서 더하는건건가요..?

File: transforms.py [hunk 2]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+    def _mcs_aug(x, y):
+        '''
+            x: (batch, time, freq, chan)
+            y: (batch, time, n_classes*4)
+        ''' 
+        batch, time, freq, chan = x.shape
+        x = tf.cast(x, tf.float64) # because of overflow and underflow
+
+        # initialize rnoisy, rnoise, phi's
+        rnoisy = tf.matmul(tf.transpose(x, [0, 2, 3, 1]), tf.transpose(x, [0, 2, 1, 3])) / time # (batch, freq, chan, chan)
```
Comment (daniel03c1) [inline transforms.py:203]: 기존 코드 보면 R_noisy 연산할 때, 그냥 elementwise square하고 num_frames로 나누는데, 
여기에 matmul이 들어간 이유가 어떻게 되나요?

Comment (daniel03c1) [reply transforms.py:203]: 아 제가 matlab 코드를 이해를 못한 것 같은데, matmul인가 보군요 확인했습니다.

File: transforms.py [hunk 3]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+        phi_noise = tf.math.real(tf.linalg.trace(tf.matmul(yyh, rnoise_inv[:,tf.newaxis,...]) / chan)) # (batch, time, freq)
+        # --------------------------------initialize end--------------------------------
+        
+
+        for it in range(iteration):
+            rnoisy_onbin = stab(rnoisy, chan, theta)
+            rnoise_onbin = stab(rnoise, chan, theta)
+
+            rnoisy_inv = tf.linalg.inv(rnoisy_onbin)
+            rnoise_inv = tf.linalg.inv(rnoise_onbin)
```
Comment (daniel03c1) [inline transforms.py:224]: 첫 iteration에는 이미 이게 구해져 있는 것 같은데, 첫 itertion에는 계산하지 않는건 어떤가요?

File: transforms.py [hunk 4]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+            lambda_noisy = safe_div(p_noisy, (p_noise + p_noisy))
+
+            phi_noise = tf.math.real(tf.linalg.trace(tf.matmul(yyh, rnoise_inv[:,tf.newaxis,...]) / chan))
+            phi_noisy = tf.math.real(tf.linalg.trace(tf.matmul(yyh, rnoisy_inv[:,tf.newaxis,...]) / chan))
+
+            rnoisy_accu = safe_div(lambda_noisy, phi_noisy)[...,tf.newaxis,tf.newaxis] * yyh
+            rnoise_accu = safe_div(lambda_noise, phi_noise)[...,tf.newaxis,tf.newaxis] * yyh
+            
+            rnoisy = safe_div(tf.reduce_sum(rnoisy_accu, axis=1) , tf.reduce_sum(lambda_noisy, axis=1)[...,tf.newaxis, tf.newaxis])
+            rnoise = safe_div(tf.reduce_sum(rnoise_accu, axis=1) , tf.reduce_sum(lambda_noise, axis=1)[...,tf.newaxis, tf.newaxis])
```
Comment (daniel03c1) [inline transforms.py:247]: 마지막 iteration에는 이 파트 필요 없는것 같은데 마지막 iteration에는 여기 작동 안하게 하는건 어떤가요?

File: transforms.py [hunk 5]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+        yx = x[..., tf.newaxis] # (batch, time, freq, chan, 1)
+        yyh = tf.matmul(yx, tf.transpose(yx, [0, 1, 2, 4, 3])) # (batch, time, freq, chan, chan)
+        rnoisy_onbin = stab(rnoisy, chan, theta)
+        rnoise_onbin = stab(rnoise, chan, theta)
+        
+        rnoisy_inv = tf.linalg.inv(rnoisy_onbin) # (batch, freq, chan, chan)
+        rnoise_inv = tf.linalg.inv(rnoise_onbin) # (batch, freq, chan, chan)
+
+        phi_noisy = tf.math.real(tf.linalg.trace(tf.matmul(yyh, rnoisy_inv[:,tf.newaxis,...]) / chan)) # (batch, time, freq)
+        phi_noise = tf.math.real(tf.linalg.trace(tf.matmul(yyh, rnoise_inv[:,tf.newaxis,...]) / chan)) # (batch, time, freq)
```
Comment (daniel03c1) [inline transforms.py:215]: matmul하고 trace구하는거면, rnoisy_inv, rnoise_inv에 마지막 두채널 transpose하고,
element wise multiplication을 yyh랑 한다음에 reduce_sum(axis=-1)하면 어떨까요?

Comment (daniel03c1) [reply transforms.py:215]: 240, 241번째 줄도 해당되는 이야기입니다!

File: transforms.py [hunk 6]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+
+def mcs_aug(iteration: int, theta = 1e-6):
+    # reference: https://github.com/funcwj/cgmm-mask-estimator.git
+    def _mcs_aug(x, y):
+        '''
+            x: (batch, time, freq, chan)
+            y: (batch, time, n_classes*4)
+        ''' 
+        batch, time, freq, chan = x.shape
+        x = tf.cast(x, tf.float64) # because of overflow and underflow
```
Comment (daniel03c1) [inline transforms.py:200]: 64로 바꾸면 연산시간과 연산량이 많아질거같은데, 반드시 필요하면 32로 방법을 찾아보는건 어떤가요?

**Comment by daniel03c1** [comment]
혹시 Matlab 코드로 돌린 결과물하고 비교는 어떻게 됐나요? wav 넣었을 때 결과물이 비슷한건지 확인 된건가요?
그리고 최적화할 때는 파이썬으로 가깝게 구현된 최적화하지 않은 코드를 기준점으로 삼아서 
최적화 함수를 별도로 만들어 비교하면서 최적화해나가는건 어떤가요?

File: transforms.py [hunk 7]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+
+        for it in range(iteration):
+            rnoisy_onbin = stab(rnoisy, chan, theta)
+            rnoise_onbin = stab(rnoise, chan, theta)
+
+            rnoisy_inv = tf.linalg.inv(rnoisy_onbin)
+            rnoise_inv = tf.linalg.inv(rnoise_onbin)
+
+            # corre = yyh
+            k_noise = tf.matmul(x[...,tf.newaxis,:], safe_div(rnoise_inv[:,tf.newaxis,...], phi_noise[...,tf.newaxis,tf.newaxis]))
```
Comment (daniel03c1) [inline transforms.py:227]: 원본 코드에서는 obs' * ... * obs 이런식으로 x가 아니라 x의 complex conjugate transpose랑 곱해야 하는데,
여기서는 그런 과정이 빠진 것 같습니다.

Comment (daniel03c1) [reply transforms.py:227]: 즉 입력이 float라도 실수부 허수부가 같이 있는 입력을 받거나, complex 형태의 입력을 이용해야하지 않을까 싶습니다.

File: transforms.py (hunk 3 above)
Comment (daniel03c1) [inline transforms.py:224]: rnoisy_inv, rnoise_inv 사용처를 보면 꼭 [:, tf.newaxis, ...]와 같이 사용되는데,
정의하는 이 파트에서 그냥 붙여주면 뒤에 코드가 조금 더 간단해지지 않을까 생각됩니다.

File: transforms.py [hunk 8]
```diff
@@ -157,3 +159,95 @@ def acs_aug(x, y):
...
+            k_noisy = tf.matmul(x[...,tf.newaxis,:], safe_div(rnoisy_inv[:,tf.newaxis,...], phi_noisy[...,tf.newaxis,tf.newaxis]))
+            k_noisy = tf.squeeze(tf.matmul(k_noisy, x[...,tf.newaxis]), axis=(-2,-1))
+            det_noisy =  tf.linalg.det(phi_noisy[...,tf.newaxis,tf.newaxis] * rnoisy_onbin[:,tf.newaxis]) * pi
+            p_noisy = tf.math.real(safe_div(tf.math.exp(-k_noisy), det_noisy)) + theta
+
+            lambda_noise = safe_div(p_noise, (p_noise + p_noisy))
+            lambda_noisy = safe_div(p_noisy, (p_noise + p_noisy))
+
+            phi_noise = tf.math.real(tf.linalg.trace(tf.matmul(yyh, rnoise_inv[:,tf.newaxis,...]) / chan))
+            phi_noisy = tf.math.real(tf.linalg.trace(tf.matmul(yyh, rnoisy_inv[:,tf.newaxis,...]) / chan))
```
Comment (daniel03c1) [inline transforms.py:241]: 사용처를 보면 [..., tf.newaxis, tf.newaxis] 이렇게 꼭 붙어있는데, 정의하는 이파트에서
한번만 하면 사용할때마다 저 문구를 붙여주지 않아도 되서 조금 더 보기 좋지 않을까 생각이 듭니다.

File: transforms.py (hunk 1 above)
Comment (lsg1213) [reply transforms.py:189]: 후자가 맞습니다 non invertible 하거나 nearly singularity하면 constant를 키워가면서 invertible하게 해주는 코드에요

**Comment by lsg1213** [comment]
> 혹시 Matlab 코드로 돌린 결과물하고 비교는 어떻게 됐나요? wav 넣었을 때 결과물이 비슷한건지 확인 된건가요?
> 그리고 최적화할 때는 파이썬으로 가깝게 구현된 최적화하지 않은 코드를 기준점으로 삼아서
> 최적화 함수를 별도로 만들어 비교하면서 최적화해나가는건 어떤가요?

개발했던 방식이 matlab코드 결과와 비교해나가면서 python numpy로 먼저 matlab 코드를 옮겼고 해당 numpy 결과와 tf를 비교해나가면서 tf로 옮겼습니다

File: transforms.py (hunk 6 above)
Comment (lsg1213) [reply transforms.py:200]: 원래 iteration이 30이어서 커가지고 터지는가 했는데 iteration을 5로 해도 터져가지고 방법을 좀 찾아봐야 될 것 같아요 최댓값을 제약해야 할지..

File: transforms.py (hunk 7 above)
Comment (lsg1213) [reply transforms.py:227]: 감사합니다. 그런데 저희는 입력자체가 허수부는 다 날아가있는 x를 받아서 어떻게 해야 할까요..? freq domain으로 transform하기 전 mic랑 foa로 mcs를 먼저 하고 해야할까요?

File: transforms.py (hunk 3 above)
Comment (lsg1213) [reply transforms.py:224]: 감사합니다.

File: transforms.py (hunk 7 above)
Comment (daniel03c1) [reply transforms.py:227]: 음... pipeline? 데이터 로드 단계부터 복소수 형태로 받도록 조금 설계를 바꿔야할까요?
1등 팀이 이것을 사용하긴 했는데, beamforming이라 하면, mic 포멧은 괜찮을 것 같긴 한데,
foa 포멧에 beam forming을 그대로 사용해도 되는지 박사님한테 물어보면 어떨까요?

File: transforms.py (hunk 6 above)
Comment (daniel03c1) [reply transforms.py:200]: 혹시 어느 부분에서 터지나요?

Comment (lsg1213) [reply transforms.py:200]: 그 부분은 정확하게 표시가 안 돼요 nan이 하나라도 생기면 거기서 바로 터지는게 아니라 invert 할 때 터져가지고요

File: transforms.py (hunk 7 above)
Comment (lsg1213) [reply transforms.py:227]: 넵 그럼 박사님께 여쭤볼게요
